### PR TITLE
bugfix: Fix issues with jars containing + on windows

### PR DESCRIPTION
**Pull request by tgodzik**
Previously, Metals would fail to read files from jars when they contained +. Now, we added a couple of fallbacks to make sure that works correctly.

Fixes https://github.com/scalameta/metals/issues/4896

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala [hunk 1]
```diff
@@ -190,12 +190,21 @@ final class FileDecoderProvider(
...
   }
 
-  private def decodeJar(uri: URI): DecoderResponse = {
+  private def decodeJar(uri: URI, original: String): DecoderResponse = {
     Try {
       val path = uri.toAbsolutePath
       FileIO.slurp(path, StandardCharsets.UTF_8)
     } match {
-      case Failure(exception) => DecoderResponse.failed(uri, exception)
+      case Failure(_) =>
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala:198]: Do we know what is thrown here if it fails? Is this just a `FileNotFound` exception? If so it _might_ be nicer to instead do like a:

```
  private def decodeJar(
      uri: URI,
      fallback: Option[String] = None,
  ): DecoderResponse = {
    Try {
      val path = uri.toAbsolutePath
      FileIO.slurp(path, StandardCharsets.UTF_8)
    } match {
      case Failure(_: FileNotFoundException) if fallback.nonEmpty =>
        decodeJar(new URI(fallback.get))
      case Failure(exception) =>
        scribe.warn(s"Unable to decode: $fallback", exception)
        DecoderResponse.failed(uri, exception)
      case Success(value) => DecoderResponse.success(uri, value)
    }
  }
```
But take it or leave it.

File: mtags/src/main/scala-2/scala/meta/internal/mtags/MtagsEnrichments.scala [hunk 2]
```diff
@@ -129,28 +130,44 @@ trait MtagsEnrichments extends CommonMtagsEnrichments {
     def stripBackticks: String = value.stripPrefix("`").stripSuffix("`")
     def isBackticked: Boolean =
       value.size > 1 && value.head == '`' && value.last == '`'
-    def toAbsolutePath: AbsolutePath = toAbsolutePath(true)
+    def toAbsolutePath: AbsolutePath = {
+      toAbsolutePath(true)
+    }
```
Comment (ckipp01) [inline mtags/src/main/scala-2/scala/meta/internal/mtags/MtagsEnrichments.scala:135]: Is this change needed? To wrap it in `{}`?

File: mtags/src/main/scala-2/scala/meta/internal/mtags/MtagsEnrichments.scala [hunk 3]
```diff
@@ -129,28 +130,44 @@ trait MtagsEnrichments extends CommonMtagsEnrichments {
...
-      else if (value.toUpperCase.startsWith("JAR")) {
+      /* Windows sometimes treats % literally, but not sure if it's always the case.
+       * https://learn.microsoft.com/en-us/troubleshoot/windows-client/networking/url-encoding-unc-paths-not-url-decoded
+       * This function tries to apply different heuristics to get the proper file system.
+       */
+      def withTryDecode(value: String)(f: String => AbsolutePath) = {
         try {
-          new URI("jar", decoded.stripPrefix("jar:"), null)
-            .toAbsolutePath(followSymlink)
+          if (Properties.isWin) f(URLDecoder.decode(value, "UTF-8"))
```
Comment (ckipp01) [inline mtags/src/main/scala-2/scala/meta/internal/mtags/MtagsEnrichments.scala:144]: Freaking windows.

File: metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala (hunk 1 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala:198]: It can be either FileNotFoundException or FilesystemNotFound, so I might use that, thanks!

File: mtags/src/main/scala-2/scala/meta/internal/mtags/MtagsEnrichments.scala (hunk 3 above)
Comment (tgodzik) [reply mtags/src/main/scala-2/scala/meta/internal/mtags/MtagsEnrichments.scala:142]: I am sort of trying all the possibilities because it seems every time one thing is fixed the other breaks, so instead let's hope fallback work.

File: metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala (hunk 1 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala]: Ok, so I ended up not doing this since the fallback uses the toAbsolutePath on String and not `new URI(fallback.get)`. String.toAbsolutePath  could work in all situations, but since this is rather fragile I will keep the old behaviour and fallback to it.

I changed it though, so it's much shorter while doing the same thing.

File: mtags/src/main/scala-2/scala/meta/internal/mtags/MtagsEnrichments.scala (hunk 2 above)
Comment (tgodzik) [reply mtags/src/main/scala-2/scala/meta/internal/mtags/MtagsEnrichments.scala]: Removed

**Review by ckipp01 — APPROVED** [review]
We keep on adding more tests around these things, so I'm confident we'll start to nail down all these places with weird windows issues 💪🏼 .

LGTM!
